plugins/toolkit/jetbrains-core/tst/software/aws/toolkits/jetbrains/services/codewhisperer/amazonqFeatureDev/controller/FeatureDevControllerTest.kt

```diff
@@ -0,0 +1,112 @@
+import software.aws.toolkits.jetbrains.services.amazonq.auth.AuthNeededStates
+import software.aws.toolkits.jetbrains.services.amazonq.messages.MessagePublisher
+import software.aws.toolkits.jetbrains.services.amazonq.toolwindow.AmazonQToolWindowFactory
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.clients.FeatureDevClient
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.controller.FeatureDevController
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.messages.IncomingFeatureDevMessage
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.session.RefinementState
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.session.Session
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.storage.ChatSessionStorage
+import software.aws.toolkits.jetbrains.services.amazonqFeatureDev.util.uploadArtifactToS3
+import software.aws.toolkits.jetbrains.services.codewhisperer.amazonqFeatureDev.FeatureDevTestBase
+
+class FeatureDevControllerTest : FeatureDevTestBase() {
+    @Rule
+    @JvmField
+    val ruleChain = RuleChain(projectRule, disposableRule)
+
+    private lateinit var controller: FeatureDevController
+    private lateinit var messanger: MessagePublisher
+    private lateinit var chatSessionStorage: ChatSessionStorage
+    private lateinit var appContext: AmazonQAppInitContext
+    private lateinit var authController: AuthController
+    private lateinit var spySession: Session
+    private lateinit var featureDevClient: FeatureDevClient
+
+    private val tabId = "tabId"
+
+    @Before
+    override fun setup() {
+        super.setup()
+        featureDevClient = mock()
+        messanger = mock()
+        chatSessionStorage = mock()
+        appContext = mock<AmazonQAppInitContext> {
+            on { project }.thenReturn(project)
+            on { messagesFromAppToUi }.thenReturn(messanger)
+        }
+        authController = spy(AuthController())
+        spySession = spy(Session("tabId", project))
+        mockkStatic("software.aws.toolkits.jetbrains.services.amazonqFeatureDev.util.UploadArtifactKt")
+        every { uploadArtifactToS3(any(), any(), any(), any(), any()) } just runs
+
+        ToolWindowManager.getInstance(projectRule.project).registerToolWindow(
+            RegisterToolWindowTask(id = AmazonQToolWindowFactory.WINDOW_ID, canWorkInDumbMode = true)
+        )
+
+        controller = FeatureDevController(appContext, chatSessionStorage, authController)
+    }
+
+    @Test
+    fun `test new tab opened`() {
+        val message: IncomingFeatureDevMessage.NewTabCreated = IncomingFeatureDevMessage.NewTabCreated("new-tab-created", tabId)
+        whenever(chatSessionStorage.getSession(any(), any())).thenReturn(spySession)
+
+        runTest {
+            controller.processNewTabCreatedMessage(message)
+        }
+        verify(authController, times(1)).getAuthNeededStates(any())
```

Comment:
suggestion: here it could be tested if the getAuthNeededStates is called with the mock project created above, rather than any input.

Reply:
Updated `getAuthNeededStates` and verified for `getSessionInfo` method being called